lib/Transforms/ListenerCSE.cpp

```diff
@@ -0,0 +1,279 @@
+//===-- ListenerCSE.cpp - Common subexpr elimination with a listener ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Transforms/ListenerCSE.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "llvm/ADT/ScopedHashTable.h"
+#include "llvm/Support/RecyclingAllocator.h"
+#include <deque>
+
+using namespace mlir;
+
+namespace {
```

Comment:
Could you please annotate which parts of this are taken from the "main" CSE verbatim and which parts are new? This will help keeping code in sync and eventually merging it.